### PR TITLE
Feature/twitter 

### PR DESCRIPTION
**Pull request by hursevich**

### CODE REVIEW

File: service/src/main/java/com/zebrunner/reporting/service/TwitterService.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.stereotype.Service;
+
+import javax.inject.Inject;
+import java.util.List;
+
+@Service
+@ConditionalOnProperty(name = "twitter.enabled", havingValue = "true")
+public class TwitterService {
+
+    private Twitter twitter;
```
Comment (brutskov) [inline service/src/main/java/com/zebrunner/reporting/service/TwitterService.java:20]: add private modificator

File: service/src/main/java/com/zebrunner/reporting/service/TwitterService.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    public TwitterService(@Value("${twitter.consumer-key}") String consumerKey,
+                          @Value("${twitter.consumer-secret}") String consumerSecret) {
+        twitter = new TwitterTemplate(consumerKey, consumerSecret);
```
Comment (brutskov) [inline service/src/main/java/com/zebrunner/reporting/service/TwitterService.java:24]: Is it possible to configure Twitter client as bean within some configuration class and autowire it?

File: web/build.gradle [hunk 3]
```diff
@@ -21,6 +21,7 @@ dependencies {
     implementation libraries.boot_starter_validation
     implementation libraries.boot_starter_actuator
     implementation libraries.spring_security_ldap
+    implementation libraries.spring_twitter
```
Comment (brutskov) [inline web/build.gradle:24]: Get rid of the dependency(use app DTO instead)

File: service/src/main/java/com/zebrunner/reporting/service/TwitterService.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public List<Tweet> getUserTimeline(String userName, int pageSize) {
```
Comment (brutskov) [inline service/src/main/java/com/zebrunner/reporting/service/TwitterService.java:27]: Create Tweet domain object with all required properties. (Tweet is third-part library domain object. It should be accessible inside this service only)

File: web/src/main/java/com/zebrunner/reporting/web/SocialController.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+@RestController
+@ConditionalOnProperty(name = "twitter.enabled", havingValue = "true")
+public class SocialController extends AbstractController implements SocialDocumentedController {
+
+    @Autowired
+    private TwitterService twitterService;
+
+    @GetMapping("/twitter/timeline")
+    @Override
+    public List<Tweet> getUserTimeline(@RequestParam("userName") String userName, @RequestParam(value = "pageSize", defaultValue = "20") int pageSize) {
```
Comment (brutskov) [inline web/src/main/java/com/zebrunner/reporting/web/SocialController.java:37]: Use domain objects and DTOs only.

File: service/src/main/java/com/zebrunner/reporting/service/TwitterService.java (hunk 2 above)
Comment (nsidorevich) [reply service/src/main/java/com/zebrunner/reporting/service/TwitterService.java:24]: Good point. Move this to configuration class, make Twitter both private and final and wire to constructor. + use @RequiredArgsConstructor

File: web/src/main/java/com/zebrunner/reporting/web/SocialController.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import javax.validation.Valid;
+import java.util.List;
+import java.util.stream.Collectors;
+
+
+@CrossOrigin
+@RequestMapping(path = "api/social", produces = MediaType.APPLICATION_JSON_VALUE)
+@RestController
+@ConditionalOnProperty(name = "twitter.enabled", havingValue = "true")
+public class SocialController extends AbstractController implements SocialDocumentedController {
```
Comment (nsidorevich) [inline web/src/main/java/com/zebrunner/reporting/web/SocialController.java:30]: Is there a reason why you extend AbstractController? Don't do that unless you have a good reason, this is something we are trying to eliminate

File: web/src/main/java/com/zebrunner/reporting/web/SocialController.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import com.zebrunner.reporting.service.TwitterService;
+import com.zebrunner.reporting.web.documented.InvitationDocumentedController;
+import com.zebrunner.reporting.web.documented.SocialDocumentedController;
+import org.dozer.Mapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.http.MediaType;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.social.twitter.api.Tweet;
+import org.springframework.web.bind.annotation.*;
```
Comment (nsidorevich) [inline web/src/main/java/com/zebrunner/reporting/web/SocialController.java:19]: Do not use import with asterisk - make them all explicit.

File: web/src/main/java/com/zebrunner/reporting/web/SocialController.java (hunk 5 above)
Comment (nsidorevich) [reply web/src/main/java/com/zebrunner/reporting/web/SocialController.java:37]: I think for this particular use case we can make an exclusion if we don't really perform any additional transformation on Tweet body. This DTO comes from spring social and we probably shouldn't create our own wrappers (not right now at least)

**Pull request merged**